### PR TITLE
Harry/small updates on archiver

### PR DESCRIPTION
**Pull request by hhlim12**
I wrote some updates on the archiving function. See `Files changed` for more detail:
1. Now the archive option specified in the yaml file is used. Currently only two keys are used, namely the name format (can be unique ID or custom string) and the file format (can be json or db file). More keywords are possible in the future.
2. Fix a bug in QE archiver:
image

Previously `calc` is not define in the if condition for QE. I think it will result in error.
3. Minor update on the printing status for parallel jobscript: print ARCHIVED to log file after finish archiving.
image

Below is the yaml and csv file I used for quick testing.

```yaml
---
###################################################################################
#Configuration file used to run high-troughput calculation using STATE (or QE) code
#v.0.1
###################################################################################

comp_spec:
     csv_loc: ./list_of_single_atoms.csv # Location of csv file
     pw_loc: /sqfs/home/u6b589/STATE/state-5.6.8/src  # Location of executable file
     pw_name: STATE
     mpi_command: "mpirun"
     pseudo_loc: /sqfs/home/u6b589/STATE/gncpp_pack   # Location of pseudopotentials directory
     n_cpus: 4                          # Number of allocated cpus for each calculation
     n_nodes: 1                          # Number of allocated node(s) for each calculation
     prefix: atom_run                         # Prefix for the directory name
     file_prefix: state
system_spec:
     type: Atom                        # One of the following: Atom, Bulk, Molecule, Surface, Adsorption
     fix_params:
        Vacuum: 10.0                   # If type is Atom, Molecule, or Surface
#        Alat: from_ase                 # If type is Bulk or surface
#        Size: (1,1,1)                  # If type is Surface
#     vary_params:

dft_spec:
    name: STATE                          # Can be STATE or QE
    fix_params:                          # List of fixed parameters for each calculation
        "KPOINT_MESH": [1, 1, 1]
        "KPOINT_SHIFT": ["OFF", "OFF", "OFF"]
        "NEG": 8
        "NSCF": 1000
        "TASK": "SCF"
        "WAY_MIX": 6
        "MIX_WHAT": 1
        "KBXMIX": 30
        "MIX_ALPHA": 0.5
        "WIDTH": -0.0020
        "XCTYPE": "ggapbe"
        "EDELTA": 0.1
    vary_params:                         # List of STATE parameters to be read from csv file
        - "PSEUDOS"
        - "GMAX"
        - "GMAXP"

archive_option:
     name_format: unique_id                     # Name of archive file.
     file_format: json                    # Type of archive file. Can be json or db

```

and csv file:
```
Species,PSEUDOS,GMAX,GMAXP
H,H@pot.H_lda1,4,16
Li,Li@pot.Li_pbe1TM,4,16
Be,Be@pot.Be_pbe1TM,4,16
B,B@pot.B_pbe3,4,16
C,C@pot.C_pbe1,4,16
```

### CODE REVIEW

File: statebatch/statebatch/archiver.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 from datetime import datetime
 from ase.db import connect
 
-def prearchive(system_type, atoms_obj, atoms_dict, comp_spec):
+def prearchive(system_type, atoms_obj, atoms_dict, comp_spec, archive_option):
```
Comment (kimrojas) [inline statebatch/statebatch/archiver.py:7]: Thanks for adding this flexibility option. However, is it possible to apply some sort of defaults parameters fallback? For example if `archive_option` is not set the `name_format` and `file_format` will still be available.

Maybe this code snippet maybe useful

```python
# 1. Set the default of archive_option input variable to be an empty dictionary
# def prearchive(system_type, atoms_obj, atoms_dict, comp_spec, archive_option={}):

# 1. Create default dictionary
archive_option_default = {"name_format": "unique_id", "file_format": "json"}

# 2. Update archive_option with archive_option_default, 
# the following code will only add default values if it does not exist in archive option
for key, value in archive_option.items():
    archive_option.setdefault(key, value)
```

More info on the syntax [here](https://note.nkmk.me/en/python-dict-setdefault/)

File: statebatch/statebatch/archiver.py [hunk 2]
```diff
@@ -76,4 +78,9 @@ def archive():
         
         # database update
         db_file.update(id = 1, atoms = atoms_obj)
-        os.rename('prerun.json', 'run.json')
\ No newline at end of file
+
+        # Rename database if unique_id is requested
+        if prerun_fname.split('.')[0] == 'unique_id':
```
Comment (kimrojas) [inline statebatch/statebatch/archiver.py:83]: This may be a little hard to follow (cause of the splitting). 
Instead can you change to the dictionary syntax ?
archive_option.get('name_format') 
archive_option.get('file_format')

**Review by kimrojas — CHANGES_REQUESTED** [review]

File: statebatch/statebatch/archiver.py (hunk 2 above)
Comment (hhlim12) [reply statebatch/statebatch/archiver.py]: I modified the code like this in which I assigned more intuitive name for the variables: 
image

I don't use `archive_option` dictionary because the only information passed to the archive() function is the name of prerun file (not the dictionary).

image

File: statebatch/statebatch/archiver.py (hunk 1 above)
Comment (hhlim12) [reply statebatch/statebatch/archiver.py]: Thanks for the nice suggestion!
I modified the `STATEbatch.py` as follow:
image

I made a design in such a way that the archiving will be always done even though the `archive_option` is not specified. So if the `archive_option` is `None`, then the dictionary will take the default value. And if one of the key is not specified, that one will also take the default value.

Comment (kimrojas) [reply statebatch/statebatch/archiver.py:7]: Thanks for the edit. I externalize the defaults setting of the archiver (in a commit I made to this branch). I just created a function that returns the user's archive_option with the default (mainly your code)

You can check the code diff here 
https://github.com/hhlim12/STATE-BATCH/compare/b5d032b..92e5d05

File: statebatch/statebatch/archiver.py (hunk 2 above)
Comment (kimrojas) [reply statebatch/statebatch/archiver.py]: Thanks. I corrected a few things. The default of archive option should not be an empty dict
```python
def prearchive(system_type, atoms_obj, atoms_dict, comp_spec, archive_option={}):
# BECOMES
def prearchive(system_type, atoms_obj, atoms_dict, comp_spec, archive_option):
```

Also I just simplified a part of the code
```python
        # prerun_fname = prerun_fname.split('.')[0]
        # prerun_ftype = prerun_ftype.split('.')[-1]
        prerun_fname, prerun_ftype = prerun_fname.split('.')
```

File: statebatch/statebatch/archiver.py (hunk 1 above)
Comment (kimrojas) [reply statebatch/statebatch/archiver.py:7]: The old coode was is still in the code but in comments. If the changes are okay with you, I can merge it already.

File: statebatch/statebatch/archiver.py (hunk 2 above)
Comment (hhlim12) [reply statebatch/statebatch/archiver.py]: Thanks!

File: statebatch/statebatch/archiver.py (hunk 1 above)
Comment (hhlim12) [reply statebatch/statebatch/archiver.py:7]: Hi Kurt, thanks for the suggestion. Even though the code is now cleaner, for me the assignment of the default parameters becomes less apparent. This is actually one of the reason I explicitly set the default in the `__init__` part of `STATEbatch.py`. Further, maybe we also need to set some default values for `comp_spec`, `dft_spec`, and `system_spec`. I planned to assign all default yaml key-values at similar location (e.g., `__init__`) so it would be more tractable. What do you think? Do you have second opinion?

Comment (kimrojas) [reply statebatch/statebatch/archiver.py:7]: I see. I got your point. Maybe my approach isn't the best for this use then. Then how about, instead of externalizing the default adder, lets just add the default adder function within the STATEbatch? Or would explicit code be better ?

Comment (hhlim12) [reply statebatch/statebatch/archiver.py:7]: I think using the default adder function is nice idea. It address the centralization of the default parameter assignment and makes the code cleaner. Let me work on it..

**Comment by hhlim12** [comment]
Hi @kimrojas , I further modified the code to include the yaml and csv reader in separate function.
I named it `control.py`. Currently I don't have any idea on specifying the default value for `comp_spec`, `dft_spec`, or `system_spec`. But shall there are new default valus need to be specified, we can specify them in this `control.py` module.

image

I also put a template for the reading task file aside from csv. Now I'm thinking to add alternative for task file (e.g., json file) to compensate the inflexibility of csv. We can put this feature also inside the `get_task` function in `control.py` module.
